rust-sdk/whirlpool/src/token.rs

```diff
@@ -333,4 +352,393 @@ mod tests {
+    #[serial]
+    async fn test_token_account_with_balance() {
+        let ctx = RpcContext::new().await;
+        let amount = 1_000_000u64;
+
+        // Create a mint
+        let mint = setup_mint(&ctx).await.unwrap();
+
+        // Create ATA with initial balance
+        let ata = setup_ata_with_amount(&ctx, mint, amount).await.unwrap();
+
+        // Now prepare instructions - should not create new instructions since account exists with sufficient balance
+        let result = prepare_token_accounts_instructions(
+            &ctx.rpc,
+            ctx.signer.pubkey(),
+            vec![TokenAccountStrategy::WithBalance(mint, amount)],
+        )
+        .await
+        .unwrap();
+
+        // Verify account state
+        let account = ctx.rpc.get_account(&ata).await.unwrap();
+        let token_account = Account::unpack(&account.data).unwrap();
+        assert_eq!(token_account.amount, amount);
+        assert_eq!(token_account.mint, mint);
+        assert_eq!(token_account.owner, ctx.signer.pubkey());
+    }
+
+    #[tokio::test]
+    #[serial]
+    async fn test_insufficient_balance() {
+        let ctx = RpcContext::new().await;
+
+        // Create a mint and token account with small balance using token.rs helpers
+        let mint = setup_mint(&ctx).await.unwrap();
+        let initial_amount = 1_000u64;
+        let _ata = setup_ata_with_amount(&ctx, mint, initial_amount)
+            .await
+            .unwrap();
+
+        // Try to prepare instructions requiring more balance
+        let required_amount = 2_000u64;
+        let result = prepare_token_accounts_instructions(
+            &ctx.rpc,
+            ctx.signer.pubkey(),
+            vec![TokenAccountStrategy::WithBalance(mint, required_amount)],
+        )
+        .await;
+
+        // Should fail due to insufficient balance
+        assert!(result.is_err());
+        assert!(result
+            .unwrap_err()
+            .to_string()
+            .contains("Insufficient balance"));
+    }
+
+    #[tokio::test]
+    #[serial]
+    async fn test_existing_token_account() {
+        let ctx = RpcContext::new().await;
+
+        // Create a mint and token account using token.rs helpers
+        let mint = setup_mint(&ctx).await.unwrap();
+        let ata = setup_ata(&ctx, mint).await.unwrap();
+
+        // Verify initial state
+        let initial_account = ctx.rpc.get_account(&ata).await.unwrap();
+        assert!(Account::unpack(&initial_account.data).is_ok());
+
+        // Try to prepare instructions for existing account
+        let result = prepare_token_accounts_instructions(
+            &ctx.rpc,
+            ctx.signer.pubkey(),
+            vec![TokenAccountStrategy::WithoutBalance(mint)],
+        )
+        .await
+        .unwrap();
+
+        // Verify account wasn't modified
+        let final_account = ctx.rpc.get_account(&ata).await.unwrap();
+        assert_eq!(initial_account.data, final_account.data);
+    }
+
+    // 3. Native Token Tests
```

Comment:
Only thing I would still add is Native Token Tests with Balance (since they wrap/unwrap tokens from SOL > wSOL)